### ISSUE TITLE
Seg detection - Mask quality

### ISSUE THREAD

**Issue by sctrueew** [ultralytics/yolov5#9955]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hi,

I've trained a seg model with Yolov5-s and yolov5-L for 300 epochs. Why are the edges rough? Not so with Dataset. After changing the image to 640, it looks like this.

![image](https://user-images.githubusercontent.com/5433346/198630199-07888beb-58a4-4b23-b247-6084cb15ea69.png)

**Comment by glenn-jocher** [comment]
@sctrueew masks are predicted at 1/4 resolution of the actual image and then upscaled. You can see this downscaling in the Proto module in the Segment() head and modify as you see fit, though higher resolution masks will also slow training and inference.

**Comment by glenn-jocher** [comment]
@sergiossm example 4x upscaling effects

**Comment by sctrueew** [comment]
@glenn-jocher Hi, thanks for your reply

Could you guide me on how I change it?

**Comment by glenn-jocher** [comment]
@sctrueew you can find the Proto module here:
https://github.com/ultralytics/yolov5/blob/49156eb1d18b6314554333c4bdae5ee3e6102992/models/common.py#L834-L845

And upsampling code in segments/predict.py:
```
                masks = process_mask(proto[i], det[:, 6:], det[:, :4], im.shape[2:], upsample=True)  # HWC
```

**Comment by kingmacth** [comment]
I'm having the same issue. How did you solve it?

**Comment by sctrueew** [comment]
I have solved it in post-processing with CV, but I need to fix it in the model.

**Comment by kingmacth** [comment]
> [glenn-jocher](/glenn-jocher)

Did you try the method provided by @glenn-jocher?
How to solve it in CV, can you guide me?
I'm solving the same issue now!

**Comment by sctrueew** [comment]
> > [glenn-jocher](/glenn-jocher)
> 
> Did you try the method provided by @glenn-jocher? How to solve it in CV, can you guide me? I'm solving the same issue now!

No, I didn't. My first step was to get the polygon's points from the mask, and then I used cv::approxPolyDP to reduce and smooth the points. currently, it's working for me

**Comment by kingmacth** [comment]
> process_mask

How can I modify if I need to predict at native resolution?

**Comment by kingmacth** [comment]
> @sctrueew you can find the Proto module here:
> 
> https://github.com/ultralytics/yolov5/blob/49156eb1d18b6314554333c4bdae5ee3e6102992/models/common.py#L834-L845
> 
> And upsampling code in segments/predict.py:
> 
> ```
>                 masks = process_mask(proto[i], det[:, 6:], det[:, :4], im.shape[2:], upsample=True)  # HWC
> ```

@glenn-jocher How can I modify if I need to predict at native resolution?

**Comment by glenn-jocher** [comment]
@kingmacth all segment predictions are already returned at native resolution by segment/predict.py

**Comment by kingmacth** [comment]
> @kingmacth all segment predictions are already returned at native resolution by segment/predict.py

I want the training to also be segmented at a higher resolution, how can I modify it?

**Comment by glenn-jocher** [comment]
@kingmacth segmentation masks start from the P3/8 grid at stride 8. They're upsampled 2x-nearest to stride 4 in Proto():
https://github.com/ultralytics/yolov5/blob/75728bb3ea99113f306280b734dedcc5d7d067b1/models/common.py#L834-L845

Then the Protos are matrix multiplied at stride 4 to become masks before they are upsampled 4x-bilinear in segment/predict.py:
https://github.com/ultralytics/yolov5/blob/75728bb3ea99113f306280b734dedcc5d7d067b1/segment/predict.py#L153

You could manually modify the Proto upsample to bilinear to see if this smooths things for you. I would try that first.

@AyushExel @Laughing-q I wonder if our upsample-nearest in Proto head is causing pixelated masks. I changed interp method from bilinear to nearest here to align with our other upsample modules, and I think for exportability, because some formats (Edge TPU?) had problems with bilinear interpolation. If this fixes the issue then its probably worth revisiting though, maybe the export issue has since been resolved.

**Comment by glenn-jocher** [comment]
@AyushExel @Laughing-q makes sense to resolve this before v7.0 segmentation release. I'll take a look at the exportability question first. I'll make a PR for bilinear, train a model, then benchmark it so it's run through all export options.

**Comment by kingmacth** [comment]
> self.upsample = nn.Upsample(scale_factor=2, mode='nearest')

Thanks for the reply.
I'll give it a try!

**Comment by Laughing-q** [comment]
@sctrueew Want some more information here.
![image](https://user-images.githubusercontent.com/61612323/200980660-0e9b2fdd-248d-42fe-9219-a9a5a929c153.png)
What's the original resolution of this picture?
Is this picture drawn in 640 resolution?
What's the image resolution of your datasets?
Thanks!

**Comment by kingmacth** [comment]
> ```python
> python segment/predict.py --retina-masks
> ```

I tried "python segment/predict.py --retina-masks" and just the output was clearer, but the mask was still jagged and there was no improvement

**Comment by sctrueew** [comment]
> @sctrueew Want some more information here. ![image](https://user-images.githubusercontent.com/61612323/200980660-0e9b2fdd-248d-42fe-9219-a9a5a929c153.png) What's the original resolution of this picture? Is this picture drawn in 640 resolution? What's the image resolution of your datasets? Thanks!
> 
> EDIT: Maybe adding `--retina-masks` option will help you?
> 
> ```python
> python segment/predict.py --retina-masks
> ```

Hi,  the image resolution is 4032x3024. No, I have drawn over the original image.  My dataset contains 18K images, most of which are 4000x3000.

**Comment by kingmacth** [comment]
Do you have a problem handling the slash of the mask when downsampling?
This picture I trained and inferred the size is 1080*1920. But there are clear steps

**Comment by Laughing-q** [comment]
@kingmacth @sctrueew ok got it.
If you're training with image-size 640, you might lose a lot pixels information while downsampling to 640 from such a high resolution. And if your objects are very thin like the white lines in your image, then the slashes can be unavoidable while using downsample operation.
Some ways you can alleviate this:
- training with higher resolution(even your original resolution if it's acceptable to you)
- modify the `Proto` module like @glenn-jocher said, you can modify it to predict masks at 1/2 resolution or 1/1 resolution.(default is 1/4 resolution)
- try some post-processing methods, like @sctrueew mentioned `cv2.approxPolyDP` which you can generate segment points first then use this cv api.

The first and second one can be used together but both of them will increase cuda memory and slow-down training and inference.

**Comment by Laughing-q** [comment]
@kingmacth  I think your prediction(this white line) might not have enough pixels in this image even through the resolution is 1080x1920, can you show me your predictions with the whole image?

**Comment by kingmacth** [comment]
> I think your prediction(this white line) might not have enough pixels in this image even through the resolution is 1080x1920, can you show me your predictions with the whole image?

This lineweight average is greater than 50 pixels. The range is between 30-100 pixels!

**Comment by sctrueew** [comment]
@kingmacth any news? Can you tell me if the problem has been resolved?

**Comment by kingmacth** [comment]
> * approxPolyDP

I use higher resolution training and inference, and the results are processed with approxPolyDP to achieve better results.

**Comment by glenn-jocher** [comment]
@sctrueew @kingmacth good news 😃! Your original issue may now be fixed ✅ in PR #10350 by @Laughing-q. This PR improves and speeds up --retina-mask mode during inference, i.e. `python segment/predict.py --retina` will plot smoother and more accurate masks at a slower plotting speed than the default masks.

To receive this update:
- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://bit.ly/yolov5-paperspace-notebook"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"></a> <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!
